org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/InvisibleProjectImporter.java

```diff
@@ -167,6 +176,136 @@ public static boolean loadInvisibleProject(IPath javaFile, IPath rootPath, boole
+	 * @param triggerFolder the folder which contains the trigger file. A search will be executed
+	 *                      based on the trigger folder.
+	 * @param linkedFolderPath the path of invisible project's linked folder.
+	 */
+	private static Collection<File> collectFoldersToSearch(IPath triggerFilePath, IFolder triggerFolder,
+			IPath linkedFolderPath) {
+		Set<File> foldersToSearch = new HashSet<>();
+		if (Objects.equals(triggerFolder.getLocation(), linkedFolderPath)) {
+			foldersToSearch.addAll(getDirectChildFolders(triggerFilePath, triggerFolder));
+		} else {
+			foldersToSearch.addAll(getSiblingFolders(triggerFolder, linkedFolderPath));
+		}
+		return foldersToSearch;
+	}
+
+	/**
+	 * Get the direct child folders of the given parent folder.
+	 * @param triggerFilePath the path of the import trigger file.
+	 * @param parentFolder the parent folder.
+	 */
+	private static Collection<File> getDirectChildFolders(IPath triggerFilePath, IFolder parentFolder) {
+		File parent = parentFolder.getLocation().toFile();
+		if (parent.isFile()) {
+			return Collections.emptySet();
+		}
+
+		Set<File> children = new HashSet<>();
+		File[] childrenFiles = parent.listFiles();
+		for (File dir : childrenFiles) {
+			if (!dir.isDirectory()) {
+				continue;
+			}
+
+			Path eclipsePath = new Path(dir.getAbsolutePath());
+			// skip ancestor folder of the trigger file.
+			if (eclipsePath.isPrefixOf(triggerFilePath)) {
+				continue;
+			}
+
+			children.add(dir);
+		}
+		return children;
+	}
+
+	/**
+	 * Get the sibling folders of the trigger folder.
+	 * @param triggerFolder the trigger folder
+	 * @param linkedFolderPath the path of invisible project's linked folder.
+	 */
+	private static Collection<File> getSiblingFolders(IFolder triggerFolder, IPath linkedFolderPath) {
+		Set<File> siblings = new HashSet<>();
+		IResource parentFolder = triggerFolder.getParent();
+		if (parentFolder == null) {
+			return Collections.emptySet();
+		}
+
+		if (!linkedFolderPath.isPrefixOf(parentFolder.getLocation())) {
+			return Collections.emptySet();
+		}
+
+		File parent = parentFolder.getLocation().toFile();
+		if (parent.isFile()) {
+			return Collections.emptySet();
+		}
+
+		File[] peerFiles = parent.listFiles();
+		for (File peerFile : peerFiles) {
+			if (peerFile.isDirectory() && !peerFile.getName().equals(triggerFolder.getName())) {
+				siblings.add(peerFile);
+			}
+		}
+		return siblings;
+	}
+
+	/**
+	 * Collect the Java files contained in the search folders. Each folder will
+	 * only be collected one Java file (if they have) at max 3 depth level.
+	 * @param searchFolders the folders to search.
+	 */
+	private static Collection<IPath> collectTriggerFiles(Collection<File> searchFolders) {
+		JavaFileDetector javaFileDetector = new JavaFileDetector();
+		for (File file : searchFolders) {
+			try {
+				Files.walkFileTree(
+					file.toPath(),
+					EnumSet.noneOf(FileVisitOption.class),
+					3 /*maxDepth*/,
```

Comment:
Only search 3 levels for the inference.

Reply:
😀 When I set the max depth to 3, my thought is: for most unmanaged folder projects, it should be not that complex that have a deep folder structure. (Like the student assignment project). Meanwhile, limit it to 3 ensures that the perf would not be impacted too much. If it comes to a project that have more than 3 level folders, we can still recover it if user opens a file inside that deep folder.

How about keeping an eye on users' feedback after it released, we can consider adding a setting if it needs?